attributioncode/validator.go

```diff
@@ -5,18 +5,30 @@ import (
 	"crypto/sha256"
 	"encoding/base64"
 	"encoding/hex"
+	"fmt"
 	"net/url"
 	"time"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+const maxUnescapedCodeLen = 600
+
 var validAttributionKeys = map[string]bool{
-	"source":   true,
-	"medium":   true,
-	"campaign": true,
-	"content":  true,
+	"source":     true,
+	"medium":     true,
+	"campaign":   true,
+	"content":    true,
+	"experiment": true,
+	"variation":  true,
+}
+
+var requiredAttributionKeys = []string{
```

Comment:
I'm not familiar with the code here, so just going on what I see here, name indicates to me that all these must be present.

Do we really require all these to be present?  At this point I don't remember what AMO passes in (or if that functionality still exists on AMO, or if this change would affect that).  Perhaps @jvillalobos can answer the AMO part.

Reply:
[For links going to AMO](https://docs.google.com/document/d/1zQddLWnKj4TyGOrEbECLlEENN0iljUjoBrGvDFz8yrI/edit) we require source and medium, and depending on the case, campaign or content.

Reply:
Ok, if we do not always include campaign or content, then we probably shouldn't have a set of required attributes here.

Reply:
I don't have all the details about the decision to make attributes required, but I do suspect it relates in part to concerns about passing web history into client telemetry (which is a data classification concern). That is primarily an issue with the "source" param, hinted at in an [origin bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1292360#c18) for this feature. Since we only allow attribution for a subset of possible sources, we have to require source.

It's also true that without a source and some other value, attribution isn't particularly useful. Source and medium are good. Campaign and content could be optional, IMO.

Reply:
> On www, we fill up some empty params with default values to ensure they can pass this validation, which does not provide any information value over simply making them optional.

Ahh, that might be reasonable then, but a code comment here explaining that would be helpful.